### PR TITLE
New feature: add Plugin event afterFindSurvey

### PR DESCRIPTION
**Pull request by TonisOrmisson**
This is a further development of a PR started here #632:

The initial idea was to allow template to be changed to participants via plugin events. Existing events did not work to effectively change the template.

### CODE REVIEW

**Comment by Shnoulle** [comment]
Just a question : maybe add some ->set ?

Minimal : `$event->set('surveyid',$this->sid)`

**Comment by Shnoulle** [comment]
Oh, and move `$this->template=Template::templateNameFilter($this->template);` after the event get.

Because : it's here to FIX : if template didn't exist anymore : set it to default. Then seems it's needed.

**Comment by TonisOrmisson** [comment]
@Shnoulle added, thanks!

**Comment by Shnoulle** [comment]
OK, we have dev metting this afternoon :) i put this in topics of dev metting :) Thanks a lot, a great pull request (IMO)

File: application/models/Survey.php [hunk 1]
```diff
@@ -247,6 +247,19 @@ public function rules()
...
     {
+        $event = new PluginEvent('afterFindSurvey');
+        $event->set('surveyid',$this->sid) ;
+        App()->getPluginManager()->dispatchEvent($event);
+        // set the attributes we allow to be fixed
+        $allowedAttributes = array( 'template','usecookie', 'allowprev',
+            'showxquestions', 'shownoanswer', 'showprogress', 'questionindex',
+            'usecaptcha', 'showgroupinfo', 'showqnumcode', 'navigationdelay');
+        foreach ($allowedAttributes as $attribute){
+            if (!is_null($event->get($attribute)))
```
Comment (olleharstedt) [inline application/models/Survey.php]: Nit picky, but `{` should be on same line as `if`. Also, space between `)` and `{` above.

File: application/models/Survey.php [hunk 2]
```diff
@@ -247,6 +247,19 @@ public function rules()
     */
     public function fixSurveyAttribute($event)
     {
+        $event = new PluginEvent('afterFindSurvey');
+        $event->set('surveyid',$this->sid) ;
```
Comment (olleharstedt) [inline application/models/Survey.php]: No space between `)` and `;`.

**Comment by olleharstedt** [comment]
:+1:

**Comment by TonisOrmisson** [comment]
Thanks to all!

**Comment by olleharstedt** [comment]
@TonisOrmisson Please also add a manual page for this event.

**Comment by TonisOrmisson** [comment]
@olleharstedt Yes, was already looking into it, but have no build number yet. Anyway, will do that.

**Comment by olleharstedt** [comment]
I guess it will be next minor version: 2.63.0.

**Comment by TonisOrmisson** [comment]
does this seem ok:
https://manual.limesurvey.org/AfterFindSurvey

**Comment by Shnoulle** [comment]
Version is OK now, not need buildnumber ? Right ?

**Comment by olleharstedt** [comment]
Yes, version is enough.

**Comment by olleharstedt** [comment]
@TonisOrmisson Great!

**Comment by TonisOrmisson** [comment]
and an initial version of a plugin that enables to set configured templates over url parameter

https://github.com/TonisOrmisson/limesurvey-url-templates

just needs a small additional LS change access Template list via api instead of referring directly to Template:
https://github.com/TonisOrmisson/LimeSurvey/commit/e7be7baec8f1015076ceb5673d864043904b6b17

**Comment by maziminke** [comment]
@TonisOrmisson Did you test the plugin with the latest Limesurvey 3.1x version?

**Comment by TonisOrmisson** [comment]
@maziminke Im not sure. I am not actively using this plugin at this point so I can not comment on the latest releases (I have a plan for it for the future). But I can check what the status is. Feel free to note any issues here
https://github.com/TonisOrmisson/limesurvey-url-templates/issues

**Comment by maziminke** [comment]
Thanks, we will test with the latest version and will fix things as needed and send you a pull request.

**Comment by TonisOrmisson** [comment]
@maziminke my first check shows there seems to be a problem with saving the json data for plugins. Seems its a Lime issue, I'll see where it lies

**Comment by TonisOrmisson** [comment]
Works until 3.15.1
broken since 3.15.2 Some issue with LS saving json type plugin settings. 
I'll find the cause and report it

**Comment by TonisOrmisson** [comment]
this one breaks it
https://github.com/LimeSurvey/LimeSurvey/commit/2f87a0fb3271579a23c7e6a9fcd7027335bb7e5f

**Pull request merged**